File: leetcode/cpp/PassedProblems/154-find-minimum-in-rotated-sorted-array-ii.cpp
/*
 * 154-find-minimum-in-rotated-sorted-array-ii.cpp

 *
 *  Created on: Jul 15, 2015
 *      Author: hanzhe
 */

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        return re(nums, 0, nums.size() - 1);
    }

    int re(vector<int>& nums, int b, int e)
    {
    	if(b >= e - 1)
    		return nums[b] > nums[e] ? nums[e] : nums[b];
    	int m = (b+e)/2;
    	if(nums[m] < nums[b])
    		return re(nums, b+1, m);
    	else if(nums[b] == nums[e])
    	{
    		if(nums[m] < nums[e])
    			return re(nums, b, m-1);
    		else if(nums[m] > nums[e])
    			return re(nums, m+1, e);
    		else
    		{
    			int r1 = re(nums, b, m);
    			int r2 = re(nums, m+1, e);
    			return r1 > r2 ? r2 : r1;
    		}
    	}

    	/* nums[b] < nums[m]*/
    	else if(nums[b] < nums[e])
    		return nums[b];
    	else return re(nums, m+1, e);
    }
};

File: leetcode/python/154-find-minimum-in-rotated-sorted-array-ii.py
#!/usr/bin/env python
# encoding: utf-8

"""
Follow up for "Find Minimum in Rotated Sorted Array":
What if duplicates are allowed?

Would this affect the run-time complexity? How and why?
Suppose a sorted array is rotated at some pivot unknown to you beforehand.

(i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

Find the minimum element.

The array may contain duplicates.
"""



# 2次过，m>=a
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        def getm(a,b):
            if a+1 >= b: return min(nums[a], nums[b])
            if nums[a] == nums[b]: return getm(a+1, b)
            if nums[a] < nums[b]: return nums[a]
            m = (a + b)/2
            if nums[m] >= nums[a]: return getm(m+1, b)
            return getm(a+1, m)
        return getm(0, len(nums))


